### DISCUSSION TITLE
Add one-time penalty to path-sections at all_pairs_dijkstra_path_length

### DISCUSSION THREAD

**Discussion by suuuehgi** [networkx/networkx#5093] [Q&A] [answered]
I have a `MultiDiGraph` and want to use `all_pairs_dijkstra_path_length` on it -- so far, so good.

But I would like to sub-classify the edges and add a one-time penalty for the usage of edges of a certain types.

Think of it as ... say public-transit bus routes. You have a one time penalty "waiting for the bus" over the default "walking" that is independent of the traveled distance.

![test](https://user-images.githubusercontent.com/7861774/134148925-2c2a98f2-91eb-47fc-b298-6dbeff8f8ca4.jpg)

But `all_pairs_dijkstra_path_length` considers just the smallest weight* between two nodes -- so never considers the bus if I add the penalty here. Clearly, I wouldn't wait, say, 5 min for a bus if it takes me 1 min of walking but on longer distances it pays off.

How would you tackle this?

Thanks a lot!

* `_weight_function(G, weight)` using [`min(attr.get(weight, 1) for attr in d.values())`](https://github.com/networkx/networkx/blob/bcc7a7e42b6e152ac57658e884f5184b33463b0b/networkx/algorithms/shortest_paths/weighted.py#L76), where d is the edge.

**Comment by boothby** [comment, marked as answer]
What I would do in this situation is clone the nodes that buses travel through, encoding state into the graph itself.  Below, I've called the route that the bus takes the AB line.

```python
G.add_edge('A', ('A', 'on the AB line'), weight=5)
G.add_edge(('A', 'on the AB line'), ('B', 'on the AB line'), weight=.1)
G.add_edge(('B', 'on the AB line'), 'B', weight=0)
```

**Reply by suuuehgi** [reply]
That is a nice idea! Thank you very much!

![test](https://user-images.githubusercontent.com/7861774/134324360-8b91ae9d-052d-45e8-88e3-86507b3b7732.jpg)
